[IDLE-241] feat: 애플 소셜 로그인 개발

0112leesy opened the pull request:
## 🧑‍💻 작업 사항

정말 작업이 늦어 죄송합니다ㅜㅜ..
주요 작업사항은 애플 소셜로그인 구현입니다!!
다만,, 작업이 늦어지게 된 이런저런 변명을 약간 첨언 하겠습니다 😅

### 1. 스프링 시큐리티와 애플 소셜로그인
애플 소셜로그인 특징은
사용자의 액세스 토큰을 발급 받기 위한 요청을 날릴 때,
애플로부터 받은 privateKey로 sign된 JWT(clientSecret)를 함께 보내야 하는 것입니다

따라서 기존에 각 플랫폼 별 소셜로그인을 공통적으로 처리해주었던
CustomOAuth2UserService의
```
OAuth2UserService<OAuth2UserRequest, OAuth2User> delegate = new DefaultOAuth2UserService();
OAuth2User oAuth2User = delegate.loadUser(userRequest);
```
코드로 사용자 정보 요청을 보낼 수 없었습니다ㅠㅡㅠ

또한, 애플은 최초 소셜로그인 시에만 사용자 정보(이메일, 닉네임 등)을 넘겨주기 때문에
이 부분도 고려해서 분기 처리를 해주어야 합니다 🥲

각종 삽질 끝에 CustomOAuth2UserService에 애플 소셜로그인까지 공통 처리할 수 없을 것 같다고 판단하여
애플 소셜로그인을 처리하는 새로운 메서드와 API를 만들게 되었습니다

### 2. AppleOAuth2UserService
애플 소셜로그인 처리를 위해서 만들게 된 새로운 서비스 클래스입니다.
주요 메서드는 단 2개!
- authenticateWithApple
  - 사용자가 애플 로그인을 성공적으로 했을때, 애플에서 redirect되는 callback API를 처리하는 메서드입니다
  - 애플에서 redirect할 때 state와 code를 넘기게 되는데, 이때 code가 중요합니다!
  - 사용자 액세스토큰을 발급받기 위한 요청을 날릴 때 아까 말씀드린 clientSecret(JWT)과 이 code도 함께 보내야하기 때문입니다
  - 따라서 request를 매개변수로 받아 code를 추출하고, clientSecret, clientId 등과 함께 애플에게 사용자 액세스토큰 발급 요청을 보냅니다.
  - 발급 받은 토큰을 파싱해서, claim 중에 sub을 가져옵니다!
  - sub은 현재 저희 서비스의 clientId와 대응됩니다. 각 플랫폼의 소셜로그인 별 사용자의 고유한 식별값이라고 생각하시면 됩니다.
  - sub과 SocialType으로 사용자를 가져옵니다.
  - 불러온 사용자의 userId를 claim에 넣고 accessToken, refreshToken을 생성하여 redirect uri의 파라미터로 전달합니다.
  - 그러나 이 사용자가 최초 소셜로그인 사용자라면, 애플로부터 사용자 정보를 전달받았을 테고, 이 사용자 정보를 회원 DB에 저장(회원가입)한 뒤 토큰을 발급합니다.
  - 과정이 정말 길지만,, 메서드를 최대한 분리하기 위해 노력했습니다 😂
  
- withdrawApple
  - [6월 30일부터 시작되는 계정 삭제 요구 사항](https://developer.apple.com/kr/news/?id=12m75xbj)
  - 위의 지침에 따라 소셜 로그인 구현 시 로그인 연동 해제 기능도 필수로 있어야 한다고 합니다
  - 따라서 회원 탈퇴 API에서 해당 회원이 애플 소셜로그인 회원이라면 애플 로그인 연동 해제를 할 수 있도록 해당 메서드를 만들었습니다!
  - 애플 소셜로그인 연동 해제 API에 clientSecret, appleRefreshToken 등을 함께 요청 보냅니다
  - 그러면 애플 소셜로그인 연동 해제가 됩니다!
  - appleRefreshToken은 위의 액세스 토큰 발급 응답에서 함께 전달받게 되는데, 이를 redis에 저장해두도록 설정했습니다.

### 3. Apple FeignClient
기존에는 RestTemplate로 Apple에다가 사용자 토큰 요청을 보냈지만,
가독성 향상과 관심사의 분리를 위해 OpenFeign으로 전환하게 되었습니다

```
@PostMapping("/auth/token")
    @Headers("Content-Type: application/x-www-form-urlencoded")
    AppleOAuth2TokenServiceResponse getAppleToken(
            @RequestParam String client_id,
            @RequestParam String client_secret,
            @RequestParam String code,
            @RequestParam String grant_type,
            @RequestParam String redirect_uri
    );

    @PostMapping("/auth/revoke")
    @Headers("Content-Type: application/x-www-form-urlencoded")
    void revokeAppleToken(
            @RequestParam String client_id,
            @RequestParam String client_secret,
            @RequestParam String token,
            @RequestParam String token_type_hint
    );
```

위와 같이 토큰 발급, 그리고 연동 해제 API 두 개를 AppleOAuth2ServiceClient에 정의해두었습니다.

### 4. AppleOAuth2UtilService
토큰 발급, 연동 해제 시 필요한 clientSecret을 만드는 역할의 유틸 클래스입니다.

원래 AppleOAuth2UserService의 private 메서드로 있었는데,,
테스트 코드 작성 시 mocking이 어려워서 유틸 클래스로 분리하게 되었습니다.

덕분애 AppleOAuth2UserService 테스트 코드는 잘 작성할 수 있었지만,
여전히 해당 유틸 클래스는 테스트 코드 작성이 어려워서 테스트 커버리지가 71.9%로 머물러있습니다 🥲

FileReader, PemReader, KeyFactory, ECPrivateKey 등의 객체 생성 때문인 것 같은데
이슈 완료가 더 늦어지면 곤란할 것 같아서 일단 이대로 배포하고 추후 이슈를 파서 해결하려고 합니다 🥹


### 5. 해당 PR의 Actions는 왜 그렇게 무수히 많이 실패하였는가
정말 면목이 없습니다 ^-^..

1. SonarCloud workflow
application.yml에 APPLE_CLIENT_ID 환경 변수가 포함되어 있는데, 이를 읽어오지 못해서 테스트가 실패했습니다.
저는 분명히 Setup Application Test Yaml 단계에서 secrets에 있는 값을 주입해줬는데 말이지요..
```
Caused by: java.lang.IllegalArgumentException: Could not resolve placeholder 'APPLE_CLIENT_ID' in value "${APPLE_CLIENT_ID}"
    	at org.springframework.util.PropertyPlaceholderHelper.parseStringValue(PropertyPlaceholderHelper.java:180) ~[spring-core-6.0.9.jar:6.0.9]
    	at org.springframework.util.PropertyPlaceholderHelper.replacePlaceholders(PropertyPlaceholderHelper.java:126) ~[spring-core-6.0.9.jar:6.0.9]
    	at org.springframework.core.env.AbstractPropertyResolver.doResolvePlaceholders(AbstractPropertyResolver.java:239) ~[spring-core-6.0.9.jar:6.0.9]
    	at org.springframework.core.env.AbstractPropertyResolver.resolveRequiredPlaceholders(AbstractPropertyResolver.java:210) ~[spring-core-6.0.9.jar:6.0.9]
    	at org.springframework.core.env.AbstractPropertyResolver.resolveNestedPlaceholders(AbstractPropertyResolver.java:230) ~[spring-core-6.0.9.jar:6.0.9]
    	at 
```

결국 application-test.yml에 임시 설정 값을 추가해줘서 해결해줬습니다.
😂 사실 정확히 어떤 문제인지 아직 파악이 되지 않아서 차차 더 공부해보아야 할 것 같습니다.


2. API Docs workflow
자꾸 애플 소셜로그인 관련 테스트 케이스 4개가 실패했습니다.
실패 원인은 거의 다 NullPointException 이었습니다.

```
> Task :compileAotTestJava
> Task :processAotTestResources
> Task :aotTestClasses
> Task :test

AppleOAuth2UserServiceTest > 애플 소셜 로그인 회원의 토큰을 만료하고 연동을 해제한다. FAILED
    kr.mybrary.userservice.authentication.domain.exception.ApplePrivateKeyReadException at AppleOAuth2UserServiceTest.java:240

AppleOAuth2UserServiceTest > 이미 가입된 애플 소셜 로그인 회원을 로그인할 때 사용자를 찾을 수 없으면 예외가 발생한다. FAILED
    java.lang.AssertionError at AppleOAuth2UserServiceTest.java:134

AppleOAuth2UserServiceTest > 최초 애플 소셜 로그인 회원을 회원 가입 후 로그인한다. FAILED
    kr.mybrary.userservice.authentication.domain.exception.ApplePrivateKeyReadException at AppleOAuth2UserServiceTest.java:173

AppleOAuth2UserServiceTest > 이미 가입된 애플 소셜 로그인 회원을 로그인한다. FAILED
    kr.mybrary.userservice.authentication.domain.exception.ApplePrivateKeyReadException at AppleOAuth2UserServiceTest.java:81
```
여러 번의 디버깅 끝에 아까 말씀드린 readPrivateKey 부분이 말썽이라는 것을 알게 되었습니다!
그렇지만 해당 메서드의 mocking 이 어려워서,, 결국 util클래스로 빼고
util 클래스 자체를 mocking 하는 것으로 해결했습니다.

이번 트러블슈팅은 저도 찜찜한 부분이 많은 지라,,
여러모로 죄송한 마음이 큽니다 😭

### 6. SonarCloud가 지적한 bug, security hotspot
- bug
```
try {
    FileReader keyReader = new FileReader(resource.getFile());
    PemReader pemReader = new PemReader(keyReader);
    PemObject pemObject = pemReader.readPemObject();
    return pemObject.getContent();
} catch (IOException e) {
    throw new ApplePrivateKeyReadException();
}
```
위의 코드에서
Use try-with-resources or close this "PemReader" in a "finally" clause.
라는 조언을 받아서
아래와 같이 변경하였습니다

```
try (FileReader keyReader = new FileReader(resource.getFile());
     PemReader pemReader = new PemReader(keyReader)) {
    PemObject pemObject = pemReader.readPemObject();
    return pemObject.getContent();
} catch (IOException e) {
    throw new ApplePrivateKeyReadException();
}
```
이를 통해 "keyReader"와 "pemReader"가 try 블록을 벗어나면 자동으로 닫히게 되어, 자원 누수를 방지하고 코드의 가독성을 향상시킬 수 있습니다.

- security hotspot
```
.nickname(appleOAuth2TokenInfo.getFullName() + RandomStringUtils.randomNumeric(5))
```
임시 닉네임의 랜덤 숫자를 생성하는 부분에서
더 안전한 메서드를 쓰라는 조언을 받아서
```
.nickname(appleOAuth2TokenInfo.getFullName() + generateSecureRandomNumeric())

private String generateSecureRandomNumeric() {
        SecureRandom secureRandom = new SecureRandom();
        int randomValue = secureRandom.nextInt(90000) + 10000;
        return String.valueOf(randomValue);
    }
```
위와 같이 SecureRandom을 활용하도록 변경해주었습니다.


<!-- 어떤 변경사항이 있는지, PR 타이틀과 동일하다면 Skip -->

<br><br>

## 🔗 링크

<!-- 관련된 대화가 이루어진 슬랙 링크 혹은 피그마 링크를 첨부. -->
<!-- - [프레이머](https://framer.com/projects/2--gSDcZoNKqU6dqCUQUYQe-53rEr?node=tQMWSSKqy)  -->

<br><br>

## 🐰 시급한 정도

🏃‍♂️ 보통 : 최대한 빠르게 리뷰 부탁드립니다.
<!-- 🚨 긴급 : 선 어프루브 후 리뷰를 부탁드립니다. (리뷰는 추후 반영)  -->
<!-- 🐢 천천히 : 급하지 않습니다. -->

<br><br>

## 📖 참고 사항

오랜 시간 기다려주셔서 정말 정말 정말 감사합니다 😭
배포 후에도 별 문제 없기를 간절히 바라고 있습니다 🥲

<!-- 공유할 내용, 레퍼런스, 추가로 발생할 것으로 예상되는 이슈, 스크린샷 등을 넣어 주세요.  -->
<!-- 변경사항이 큰 경우 집중해야 할 부분, 또는 불안해서 봐주었으면 하는 부분 등 -->

0112leesy committed:
[IDLE-241] feat: AppleOAuth2UserInfo 생성

 - 애플 소셜로그인 시 attribute에서 각 프로필 항목을 가져오는 메서드 구현

0112leesy committed:
[IDLE-241] feat: OAuthAttributes SocialType.APPLE 추가

 - 애플 소셜로그인 시 사용자 이름은 임의로 설정

0112leesy committed:
[IDLE-241] feat: AppleController 생성

 - 애플 소셜로그인 시 리다이렉트되는 사용자 정보 값을 처리하는 컨트롤러 생성

0112leesy committed:
[IDLE-241] feat: Apple 소셜로그인 관련 DTO 생성

 - AppleDTO: 사용자 식별 및 사용자 정보 저장에 필요한 필드를 담은 DTO
 - AppleUser: 애플에서 전송되는 payload의 user 값 json 매핑을 위한 DTO

0112leesy committed:
[IDLE-241] feat: Apple 소셜로그인 및 회원 탈퇴 기능 구현

 - AppleService: 소셜로그인 및 회원 탈퇴 로직
 - AppleUtil: 토큰 API에 전송하기 위한 client secret 생성 로직
 - RedisUtil: duration이 null일 때 추가 (ttl 설정 X)
 - UserServiceImpl: 회원 탈퇴 시 애플로 로그인된 사용자라면 애플 계정 연동 해지

0112leesy committed:
[IDLE-241] refactor: 불필요한 코드 삭제

 - AppleUtil: decodeJwtTokenPayload 메서드 삭제

0112leesy committed:
[IDLE-241] feat: 의존성 추가

 - client secret 생성을 위한 nimbus-jose-jwt, bcpkix-jdk15on 추가

0112leesy committed:
[IDLE-241] refactor: 패키지 구조 개선

 - 애플 소셜로그인 콜백 api를 AuthenticationController로 옮기고 기존 AppleController 삭제
 - DTO명 변경: AppleDTO -> AppleOAuth2TokenInfo, AppleUser -> AppleOAuth2UserInfo
 - AppleUtil 메서드 AppleOAuth2UserService 내에 프라이빗 메서드로 이동
 - 서비스명 변경: AppleService -> AppleOAuth2UserService
 - 애플 소셜로그인은 스프링 시큐리티를 활용하지 않기 때문에 OAuthAttributes에서 애플 사용자 처리 로직 삭제

0112leesy committed:
[IDLE-241] refactor: AppleOAuth2UserService 상수 처리

0112leesy committed:
[IDLE-241] refactor: 애플 소셜로그인 관련 전역 상수 클래스 OAuth2Parameter 생성

0112leesy committed:
[IDLE-241] refactor: AppleOAuth2UserService 토큰 생성/회수 API RestTemplate -> FeignClient로 대체

  - 코드 가독성 향상, 관심사의 분리, 예외 처리 용이성 등의 이유로 대체
  - AppleOAuth2ServiceClient 인터페이스 생성

0112leesy committed:
[IDLE-241] feat: 애플 소셜로그인 시 예외 클래스 생성

  - AppleUserInfoNotFoundException
  - AppleCodeNotFoundException

0112leesy committed:
[IDLE-241] feat: 애플 소셜로그인 시 예외 클래스 추가

  - AppleClientSecretNotCreatedException
  - ApplePrivateKeyReadException
  - AppleTokenParseException
  - AppleUserInfoReadException
  - AppleUserNotFoundException

0112leesy committed:
[IDLE-241] refactor: 애플 소셜로그인 처리 메서드 리팩터링

  - 메서드 추출 기법을 통해 책임 분리
  - 최초 소셜 로그인 시 회원 가입 처리, 그렇지 않은 경우 로그인만 처리되도록 개선
  - 예외 처리 개선

0112leesy committed:
[IDLE-241] test: 애플 소셜 로그인 처리 API 테스트 코드 작성

  - RedisUtilTest 테스트 케이스 추가: 만료 시간이 있는 경우

0112leesy committed:
[IDLE-241] test: 애플 소셜 로그인 회원 탈퇴 테스트 케이스 추가

  - UserFixture에 APPLE_USER 추가

0112leesy committed:
[IDLE-241] test: 애플 소셜 로그인 및 연동 해제 테스트 코드 작성

  - fix: AppleUserInfoReadException 상태 코드 수정

0112leesy committed:
[IDLE-241] feat: 애플 소셜로그인 관련 설정값 추가

0112leesy committed:
[IDLE-241] feat: workflow 수정

 - Apple Auth Private Key 파일 주입 및 secret 변수 추가

0112leesy committed:
[IDLE-241] fix: workflow echo 명령어 수정

 - Apple Auth Private Key 파일 문자열의 줄바꿈 이슈 해결

0112leesy committed:
[IDLE-241] fix: workflow mkdir 명령어 추가

 - Apple Auth Private Key 파일 생성 시 폴더를 찾을 수 없는 이슈 해결

0112leesy committed:
Merge b864a7200c02705d23ca3fea6c681d47f6bdd4f6 into 224b987f1074ee279a09c8e7df0c7f7ab25908ab

0112leesy committed:
[IDLE-241] fix: @Test 어노테이션 추가

0112leesy committed:
[IDLE-241] fix: workflow 및 AppleOAuth2UserServiceTest 테스트 코드 수정

0112leesy committed:
[IDLE-241] fix: application-test.yml 수정

  - apple 소셜 로그인 관련 테스트 프로퍼티 설정

0112leesy committed:
[IDLE-241] fix: workflow 수정

  - Apple PrivateKey 파일 경로 수정

0112leesy committed:
[IDLE-241] refactor: AppleOAuth2UtilService 생성

  - clientSecret 생성 메서드 분리

0112leesy committed:
[IDLE-241] fix: sonarcloud bug, security hotspot 보완

  - 닉네임 설정 시 난수 생성 메서드 개선
  - try-with-resources 문으로 자원 누수 방지

0112leesy committed:
[IDLE-241] fix: sonarcloud code smell 보완

  - final static -> static final
  - private 생성자 추가
  - 일반 예외 대신 전용 예외 지정

0112leesy committed:
[IDLE-241] test: AppleOAuth2UtilServiceTest 추가

0112leesy committed:
[IDLE-241] fix: AppleOAuth2UtilServiceTest 제거

  - 테스트 환경에서 applePrivateKey resource를 읽는 데에 문제가 생겨 테스트 실패
  - readPrivateKey 이슈 추후 해결 예정

minnseong commented on user-service/src/main/java/kr/mybrary/userservice/authentication/domain/oauth2/service/AppleOAuth2UserService.java:
```diff
@@ -0,0 +1,173 @@
+        User findUser = userRepository.findBySocialTypeAndSocialId(SocialType.APPLE, socialId).orElse(null);
+        if (findUser == null) {
+            throw new AppleUserNotFoundException();
+        }
```
r2: Optional의 orElseThrow 문법을 사용하면 if 절을 사용하지 않고, null 처리를 할 수 있을 것 같습니다! 바로 리턴도 가능할 것 같습니다!

``` java
return userRepository.findBySocialTypeAndSocialId(SocialType.APPLE, socialId)
    .orElseThrow(AppleUserNotFoundException::new);
```

0112leesy replied:
orElseThrow 문을 생각하지 못했네요 😂 좋은 의견 감사합니다!!

minnseong commented on user-service/src/main/java/kr/mybrary/userservice/authentication/domain/oauth2/service/AppleOAuth2UserService.java:
```diff
@@ -0,0 +1,173 @@
+
+    private User signUpAndGetAppleUserFromRequest(HttpServletRequest request) {
+        AppleOAuth2UserInfo appleOAuth2UserInfo = getAppleOAuth2UserInfo(request);
+        if (request.getParameter(CODE) == null) {
```
r2 : CODE를 확인 하는 로직이 반복되는 것 같습니다. 메서드로 추출하면 좋을 것 같습니다!

``` java
if (request.getParameter(CODE) == null) {
    throw new AppleCodeNotFoundException();
}
```

0112leesy replied:
넵넵! 메서드로 추출해서 재사용하겠습니다!

minnseong commented on user-service/src/main/java/kr/mybrary/userservice/authentication/domain/oauth2/service/AppleOAuth2UserService.java:
```diff
@@ -0,0 +1,173 @@
+                .introduction("")
+                .role(Role.USER)
+                .build();
+        createdUser.updatePassword(passwordEncoder.encode(createdUser.getPassword()));
```
r3 : 빌더로 createdUser를 생성할 때, 인코딩도 같이할 수 있을 것 같습니다!
``` java
 User createdUser = User.builder()
                .socialType(SocialType.APPLE)
                .socialId(appleOAuth2TokenInfo.getId())
                .loginId(UUID.randomUUID().toString())
                .password(passwordEncoder.encode(UUID.randomUUID().toString()))
...
```

0112leesy replied:
좋은 의견 감사합니다!! 😄 의견 주신대로 수정하겠습니닷!

minnseong commented on user-service/src/main/java/kr/mybrary/userservice/authentication/domain/oauth2/service/AppleOAuth2UserService.java:
```diff
@@ -0,0 +1,173 @@
+@Service
+@Slf4j
+@Setter
+public class AppleOAuth2UserService {
```
r1 : @Transactional이 빠진 것 같습니다. 실제 userRepository를 사용하는 로직이 있기 때문에 @Transactional을 사용해야할 것 같습니다!

0112leesy replied:
userRepository로 회원 정보를 저장한다는 사실을 깜빡했었네요 @Transactional 어노테이션 추가하도록 하겠습니다!

minnseong commented on user-service/src/main/java/kr/mybrary/userservice/authentication/domain/oauth2/service/AppleOAuth2UserService.java:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    private String generateSecureRandomNumeric() {
+        SecureRandom secureRandom = new SecureRandom();
```
r3: SecureRandom을 매번 생성하는 것 보다는 전역 변수로 선언하여, 최초에 한번만 생성하고 사용하는 방법은 어떤가요?

<참고>
https://stackoverflow.com/questions/60315622/should-securerandom-be-used-as-singleton-or-a-new-object-should-be-created-each

0112leesy replied:
좋은 레퍼런스 감사합니다!
```
private static final SecureRandom secureRandom = new SecureRandom();
```
이렇게 static 변수로 선언하는 것으로 수정하겠습니다!

minnseong commented on user-service/src/main/java/kr/mybrary/userservice/authentication/domain/oauth2/service/AppleOAuth2UtilService.java:
```diff
@@ -0,0 +1,75 @@
+public class AppleOAuth2UtilService {
+
+    public String createAppleClientSecret(String clientId, String clientSecret) {
+        String[] secretKeyResourceArr = clientSecret.split("/");
```
r3: 배열 인덱스 문제가 발생할 수도 있을 것 같습니다! 예외 처리를 하는 코드를 작성하면 어떨까요? 
ex) secretKeyResourceArr의 크기를 검증하는 로직

0112leesy replied:
넵넵! clientSecret.split("/") 했을 때 크기가 3이 아니라면 예외가 발생하도록 수정하겠습니다!

minnseong commented on user-service/src/main/java/kr/mybrary/userservice/client/apple/api/AppleOAuth2ServiceClient.java:
```diff
@@ -0,0 +1,31 @@
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+
+@FeignClient(name = "appleOAuth2Client")
```
config repository에 userservice-prod.yml에 잊지 않고, 경로를 추가해놓으면 좋을 것 같습니다!

0112leesy replied:
네넵! appleOAuth2Client url 및 spring security 관련 설정 값 config repository에 추가해두겠습니다 😊

minnseong reviewed (commented):
너무너무너무너무 수고많으셨습니다!!! 배포했을 때 잘 동작하면 좋겠습니다!!
pr 메시지도 자세히 써주시고 코드도 가독성 있게 작성해주셔서 애플 소셜 로그인에 대해서 많이 알게 된 계기였습니다!

0112leesy committed:
[IDLE-241] refactor: 코드 리뷰 반영

  - r2: Optional의 orElseThrow 문법 적용
  - r2: request로부터 CODE를 확인하는 메서드 추출
  - r3: 빌더로 createdUser 생성 시 패스워드 인코딩 추가
  - r1: 서비스 클래스에 @transactional 추가
  - r3: SecureRandom 전역 변수로 선언
  - r3: secretKeyResourceArr의 크기를 검증하는 로직 추가
